README.md
# ![PHP UPS API](https://rawgit.com/gabrielbull/php-ups-api/develop/php-ups-api-logo.svg "PHP UPS API")

![CI](https://github.com/gabrielbull/php-ups-api/workflows/CI/badge.svg?branch=master)
[![StyleCI](https://styleci.io/repos/7774788/shield)](https://styleci.io/repos/7774788)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/gabrielbull/php-ups-api/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/gabrielbull/php-ups-api/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/gabrielbull/php-ups-api/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/gabrielbull/php-ups-api/?branch=master)
[![Code Climate](https://codeclimate.com/github/gabrielbull/php-ups-api/badges/gpa.svg)](https://codeclimate.com/github/gabrielbull/php-ups-api)
[![Latest Stable Version](http://img.shields.io/packagist/v/gabrielbull/ups-api.svg?style=flat)](https://packagist.org/packages/gabrielbull/ups-api)
[![Total Downloads](https://img.shields.io/packagist/dt/gabrielbull/ups-api.svg?style=flat)](https://packagist.org/packages/gabrielbull/ups-api)
[![License](https://img.shields.io/packagist/l/gabrielbull/ups-api.svg?style=flat)](https://packagist.org/packages/gabrielbull/ups-api)
[![Join the chat at https://gitter.im/gabrielbull/php-ups-api](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/gabrielbull/php-ups-api?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

This library wrap all the UPS API into a simple to use PHP Library. It currently covers the Quantum View®,
Tracking API, Shipping API, Rating API and Time in Transit API. Feel free to contribute.

## Table Of Content

1. [Requirements](#requirements)
2. [Installation](#installation)
3. [Address Validation Class](#addressvalidation-class)
    * [Example](#addressvalidation-class-example)
    * [Parameters](#addressvalidation-class-parameters)
4. [Simple Address Validation Class](#simple-addressvalidation-class)
    * [Example](#simple-addressvalidation-class-example)
    * [Parameters](#simple-addressvalidation-class-parameters)    
5. [QuantumView Class](#quantumview-class)
    * [Example](#quantumview-class-example)
    * [Parameters](#quantumview-class-parameters)
6. [Tracking Class](#tracking-class)
    * [Example](#tracking-class-example)
    * [Parameters](#tracking-class-parameters)
7. [Rate Class](#rate-class)
    * [Example](#rate-class-example)
    * [Parameters](#rate-class-parameters)
8. [RateTimeInTransit Class](#ratetimeintransit-class)
    * [Example](#ratetimeintransit-class-example)
    * [Parameters](#ratetimeintransit-class-parameters)
9. [TimeInTransit Class](#timeintransit-class)
    * [Example](#timeintransit-class-example)
    * [Parameters](#timeintransit-class-parameters)
10. [Locator Class](#locator-class)
    * [Example](#locator-class-example)
    * [Parameters](#locator-class-parameters)
11. [Tradeability Class](#tradeability-class)
    * [Example](#tradeability-class-example)
    * [Parameters](#tradeability-class-parameters)
12. [Shipping Class](#shipping-class)
    * [Example](#shipping-class-example)
    * [Parameters](#shipping-class-parameters)
13. [Pickup Class](#pickup-class)
    * [Example](#pickup-class-example)  
14. [Logging](#logging)
15. [License](#license-section)

<a name="requirements"></a>
## Requirements

This library uses PHP 7.1+.

To use the UPS API, you have to [request an access key from UPS](https://www.ups.com/upsdeveloperkit). For every request,
you will have to provide the Access Key, your UPS User ID and Password.

<a name="installation"></a>
## Installation

It is recommended that you install the PHP UPS API library [through composer](http://getcomposer.org/). To do so,
run the Composer command to install the latest stable version of PHP UPS API:

```bash
$ composer require gabrielbull/ups-api
```

If not using composer, you must also include these libraries: [Guzzle](https://github.com/guzzle/guzzle), [Guzzle Promises](https://github.com/guzzle/promises), [Guzzle PSR7] (https://github.com/guzzle/psr7), [PHP-Fig PSR Log](https://github.com/php-fig/log), and [PHP-Fig HTTP Message](https://github.com/php-fig/http-message).

<a name="addressvalidation-class"></a>
## Address Validation Class (Street Level)

The Address Validation Class allow you to validate an address at street level. Suggestions are given when address is invalid.

Note: UPS has two Address Validations. This is Street Level option, which includes all option
of the normal Address Validation class and adds street level validation.

Currently, only US & Puerto Rico are supported.

<a name="addressvalidation-class-example"></a>
### Example

```php
$address = new \Ups\Entity\Address();
$address->setAttentionName('Test Test');
$address->setBuildingName('Test');
$address->setAddressLine1('Address Line 1');
$address->setAddressLine2('Address Line 2');
$address->setAddressLine3('Address Line 3');
$address->setStateProvinceCode('NY');
$address->setCity('New York');
$address->setCountryCode('US');
$address->setPostalCode('10000');

$xav = new \Ups\AddressValidation($accessKey, $userId, $password);
$xav->activateReturnObjectOnValidate(); //This is optional
try {
    $response = $xav->validate($address, $requestOption = \Ups\AddressValidation::REQUEST_OPTION_ADDRESS_VALIDATION, $maxSuggestion = 15);
} catch (Exception $e) {
    var_dump($e);
}
```
#### AddressValidation::validateReturnAVObject()
In the code above `$xav->activateReturnObjectOnValidate()` is completely optional. Calling this method will cause 
`AddressValidation::validate()` to return an `AddressValidationResponse` object. If you do not call this method, `validate`
continues to function as it has previously. If you do not call this method, a single object with either the matched 
validated address, or the first candidate address if the address is ambiguous, will be returned.

The AddressValidationResponse object provides a number of methods to allow you to more easily query the API response to 
determine the outcome. Continuing the example from above, returning an `AddressValidationResponse` object will allow
you to be a bit more specific with how you handle the various outcomes:

```PHP
if ($response->noCandidates()) {
    //Do something clever and helpful to let the use know the address is invalid
}
if ($response->isAmbiguous()) {
    $candidateAddresses = $response->getCandidateAddressList();
    foreach($candidateAddresses as $address) {
        //Present user with list of candidate addresses so they can pick the correct one        
    }
}
if ($response->isValid()) {
    $validAddress = $response->getValidatedAddress();
    
    //Show user validated address or update their address with the 'official' address
    //Or do something else helpful...
}
```

<a name="addressvalidation-class-parameters"></a>
### Parameters

Address Validation parameters are:

 * `address` Address object as constructed in example
 * `requestOption` One of the three request options. See documentation. Default = Address Validation.
 * `maxSuggestion` Maximum number of suggestions to be returned. Max = 50
 
<a name="simple-addressvalidation-class"></a>
## Simple Address Validation Class 

The Simple Address Validation Class allow you to validate less extensive as the previous class. It returns a quality score of the supplied address and provides alternatives.

Note: UPS has two Address Validations. This is the Simple option. 

Currently, only US & Puerto Rico are supported.

<a name="simple-addressvalidation-class-example"></a>
### Example

```php
$address = new \Ups\Entity\Address();
$address->setStateProvinceCode('NY');
$address->setCity('New York');
$address->setCountryCode('US');
$address->setPostalCode('10000');

$av = new \Ups\SimpleAddressValidation($accessKey, $userId, $password);
try {
    $response = $av->validate($address);
    var_dump($response);
} catch (Exception $e) {
    var_dump($e);
}
```

<a name="simpleaddressvalidation-class-parameters"></a>
### Parameters

Simple Address Validation parameters are:

* `address` Address object as constructed in example

<a name="quantumview-class"></a>
## QuantumView Class

The QuantumView Class allow you to request a Quantum View Data subscription.

<a name="quantumview-class-example"></a>
### Example

```php
$quantumView = new Ups\QuantumView($accessKey, $userId, $password);

try {
	// Get the subscription for all events for the last hour
	$events = $quantumView->getSubscription(null, (time() - 3600));

	foreach($events as $event) {
		// Your code here
		echo $event->Type;
	}

} catch (Exception $e) {
	var_dump($e);
}
```

<a name="quantumview-class-parameters"></a>
### Parameters

QuantumView parameters are:

 * `name` Name of subscription requested by user. If _null_, all events will be returned.
 * `beginDateTime` Beginning date time for the retrieval criteria of the subscriptions. Format: Y-m-d H:i:s or Unix timestamp.
 * `endDateTime` Ending date time for the retrieval criteria of the subscriptions. Format: Y-m-d H:i:s or Unix timestamp.
 * `fileName` File name of specific subscription requested by user.
 * `bookmark` Bookmarks the file for next retrieval.

_If you provide a `beginDateTime`, but no `endDateTime`, the `endDateTime` will default to the current date time._

_To use the `fileName` parameter, do not provide a `beginDateTime`._

<a name="tracking-class"></a>
## Tracking Class

The Tracking Class allow you to track a shipment using the UPS Tracking API.

<a name="tracking-class-example"></a>
### Example using Tracking Number / Mail Innovations tracking number

```php
$tracking = new Ups\Tracking($accessKey, $userId, $password);

try {
	$shipment = $tracking->track('TRACKING NUMBER');

	foreach($shipment->Package->Activity as $activity) {
		var_dump($activity);
	}

} catch (Exception $e) {
	var_dump($e);
}
```

<a name="tracking-class-parameters"></a>
### Parameters

Tracking parameters are:

 * `trackingNumber` The package’s tracking number.
 * `requestOption` Optional processing. For Mail Innovations the only valid options are Last Activity and All activity.

<a name="tracking-class-example"></a>
### Example using Reference Number

```php
$tracking = new Ups\Tracking($accessKey, $userId, $password);

try {
    $shipment = $tracking->trackByReference('REFERENCE NUMBER');

    foreach($shipment->Package->Activity as $activity) {
        var_dump($activity);
    }

} catch (Exception $e) {
    var_dump($e);
}
```
<a name="tracking-class-parameters"></a>
### Parameters

Tracking parameters are:

 * `referenceNumber` The ability to track any UPS package or shipment by reference number. Reference numbers can be a purchase order number, job number, etc. Reference Number is supplied when generating a shipment.
 * `requestOption` Optional processing. For Mail Innovations the only valid options are Last Activity and All activity.

<a name="tracking-class-example"></a>
### Example using Reference Number with additional parameters

```php
$tracking = new Ups\Tracking($accessKey, $userId, $password);

$tracking->setShipperNumber('SHIPPER NUMBER');

$beginDate = new \DateTime('2016-01-01');
$endDate = new \DateTime('2016-01-31');

$tracking->setBeginDate($beginDate);
$tracking->setEndDate($endDate);

try {
    $shipment = $tracking->trackByReference('REFERENCE NUMBER');

    foreach($shipment->Package->Activity as $activity) {
        var_dump($activity);
    }

} catch (Exception $e) {
    var_dump($e);
}
```

The parameters shipperNumber, beginDate and endDate are optional. Either of the parameters can be set individually. These parameters can help to narrow the search field when tracking by reference, since it might happen that the reference number used is not unique. When using tracking by tracking number these parameters are not needed since the tracking number is unique.

<a name="rate-class"></a>
## Rate Class

The Rate Class allow you to get shipment rates using the UPS Rate API.

<a name="rate-class-example"></a>
### Example

```php
$rate = new Ups\Rate(
	$accessKey,
	$userId,
	$password
);

try {
    $shipment = new \Ups\Entity\Shipment();

    $shipperAddress = $shipment->getShipper()->getAddress();
    $shipperAddress->setPostalCode('99205');

    $address = new \Ups\Entity\Address();
    $address->setPostalCode('99205');
    $shipFrom = new \Ups\Entity\ShipFrom();
    $shipFrom->setAddress($address);

    $shipment->setShipFrom($shipFrom);

    $shipTo = $shipment->getShipTo();
    $shipTo->setCompanyName('Test Ship To');
    $shipToAddress = $shipTo->getAddress();
    $shipToAddress->setPostalCode('99205');

    $package = new \Ups\Entity\Package();
    $package->getPackagingType()->setCode(\Ups\Entity\PackagingType::PT_PACKAGE);
    $package->getPackageWeight()->setWeight(10);
    
    // if you need this (depends of the shipper country)
    $weightUnit = new \Ups\Entity\UnitOfMeasurement;
    $weightUnit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_KGS);
    $package->getPackageWeight()->setUnitOfMeasurement($weightUnit);

    $dimensions = new \Ups\Entity\Dimensions();
    $dimensions->setHeight(10);
    $dimensions->setWidth(10);
    $dimensions->setLength(10);

    $unit = new \Ups\Entity\UnitOfMeasurement;
    $unit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_IN);

    $dimensions->setUnitOfMeasurement($unit);
    $package->setDimensions($dimensions);

    $shipment->addPackage($package);

    var_dump($rate->getRate($shipment));
} catch (Exception $e) {
    var_dump($e);
}
```
<a name="rate-class-parameters"></a>
### Parameters

 * `rateRequest` Mandatory. rateRequest Object with shipment details

This Rate class is not finished yet! Parameter should be added when it will be finished.

<a name="ratetimeinstransit-class"></a>
## RateTimeInTransit Class

The RateTimeInTransit Class allow you to get shipment rates like the Rate Class, but the response will also include 
TimeInTransit data.

<a name="ratetimeintransit-class-example"></a>
### Example

```php
$rate = new Ups\RateTimeInTransit(
	$accessKey,
	$userId,
	$password
);

try {
    $shipment = new \Ups\Entity\Shipment();

    $shipperAddress = $shipment->getShipper()->getAddress();
    $shipperAddress->setPostalCode('99205');

    $address = new \Ups\Entity\Address();
    $address->setPostalCode('99205');
    $shipFrom = new \Ups\Entity\ShipFrom();
    $shipFrom->setAddress($address);

    $shipment->setShipFrom($shipFrom);

    $shipTo = $shipment->getShipTo();
    $shipTo->setCompanyName('Test Ship To');
    $shipToAddress = $shipTo->getAddress();
    $shipToAddress->setPostalCode('99205');

    $package = new \Ups\Entity\Package();
    $package->getPackagingType()->setCode(\Ups\Entity\PackagingType::PT_PACKAGE);
    $package->getPackageWeight()->setWeight(10);
    
    // if you need this (depends of the shipper country)
    $weightUnit = new \Ups\Entity\UnitOfMeasurement;
    $weightUnit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_KGS);
    $package->getPackageWeight()->setUnitOfMeasurement($weightUnit);

    $dimensions = new \Ups\Entity\Dimensions();
    $dimensions->setHeight(10);
    $dimensions->setWidth(10);
    $dimensions->setLength(10);

    $unit = new \Ups\Entity\UnitOfMeasurement;
    $unit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_IN);

    $dimensions->setUnitOfMeasurement($unit);
    $package->setDimensions($dimensions);

    $shipment->addPackage($package);

    $deliveryTimeInformation = new \Ups\Entity\DeliveryTimeInformation();
    $deliveryTimeInformation->setPackageBillType(\Ups\Entity\DeliveryTimeInformation::PBT_NON_DOCUMENT);
    
    $pickup = new \Ups\Entity\Pickup();
    $pickup->setDate("20170520");
    $pickup->setTime("160000");
    $shipment->setDeliveryTimeInformation($deliveryTimeInformation);

    var_dump($rate->shopRatesTimeInTransit($shipment));
} catch (Exception $e) {
    var_dump($e);
}
```
<a name="ratetimeintransit-class-parameters"></a>
### Parameters

 * `rateRequest` Mandatory. rateRequest Object with shipment details

This RateTimeInTransit extends the Rate class which is not finished yet! Parameter should be added when it will be finished.

<a name="timeintransit-class"></a>
## TimeInTransit Class

The TimeInTransit Class allow you to get all transit times using the UPS TimeInTransit API.

<a name="timeintransit-class-example"></a>
### Example

```php
$timeInTransit = new Ups\TimeInTransit($access, $userid, $passwd);

try {
    $request = new \Ups\Entity\TimeInTransitRequest;

    // Addresses
    $from = new \Ups\Entity\AddressArtifactFormat;
    $from->setPoliticalDivision3('Amsterdam');
    $from->setPostcodePrimaryLow('1000AA');
    $from->setCountryCode('NL');
    $request->setTransitFrom($from);

    $to = new \Ups\Entity\AddressArtifactFormat;
    $to->setPoliticalDivision3('Amsterdam');
    $to->setPostcodePrimaryLow('1000AA');
    $to->setCountryCode('NL');
    $request->setTransitTo($to);

    // Weight
    $shipmentWeight = new \Ups\Entity\ShipmentWeight;
    $shipmentWeight->setWeight($totalWeight);
    $unit = new \Ups\Entity\UnitOfMeasurement;
    $unit->setCode(\Ups\Entity\UnitOfMeasurement::UOM_KGS);
    $shipmentWeight->setUnitOfMeasurement($unit);
    $request->setShipmentWeight($shipmentWeight);

    // Packages
    $request->setTotalPackagesInShipment(2);

    // InvoiceLines
    $invoiceLineTotal = new \Ups\Entity\InvoiceLineTotal;
    $invoiceLineTotal->setMonetaryValue(100.00);
    $invoiceLineTotal->setCurrencyCode('EUR');
    $request->setInvoiceLineTotal($invoiceLineTotal);

    // Pickup date
    $request->setPickupDate(new DateTime);

    // Get data
    $times = $timeInTransit->getTimeInTransit($request);

	foreach($times->ServiceSummary as $serviceSummary) {
		var_dump($serviceSummary);
	}

} catch (Exception $e) {
    var_dump($e);
}
```

<a name="timeintransit-class-parameters"></a>
### Parameters

 * `timeInTransitRequest` Mandatory. timeInTransitRequest Object with shipment details, see example above.

<a name="locator-class"></a>
## Locator Class

The Locator class allows you to search for UPS Access Point locations.

<a name="locator-class-example"></a>
### Example

```php
$locatorRequest = new \Ups\Entity\LocatorRequest;

$originAddress = new \Ups\Entity\OriginAddress;
$address = new \Ups\Entity\AddressKeyFormat;
$address->setCountryCode('NL');
$originAddress->setAddressKeyFormat($address);

$geocode = new \Ups\Entity\GeoCode;
$geocode->setLatitude(52.0000);
$geocode->setLongitude(4.0000);
$originAddress->setGeoCode($geocode);
$locatorRequest->setOriginAddress($originAddress);

$translate = new \Ups\Entity\Translate;
$translate->setLanguageCode('ENG');
$locatorRequest->setTranslate($translate);

$acccessPointSearch = new \Ups\Entity\AccessPointSearch;
$acccessPointSearch->setAccessPointStatus(\Ups\Entity\AccessPointSearch::STATUS_ACTIVE_AVAILABLE);

$locationSearch = new \Ups\Entity\LocationSearchCriteria;
$locationSearch->setAccessPointSearch($acccessPointSearch);
$locationSearch->setMaximumListSize(25);

$locatorRequest->setLocationSearchCriteria($locationSearch);

$unitOfMeasurement = new \Ups\Entity\UnitOfMeasurement;
$unitOfMeasurement->setCode(\Ups\Entity\UnitOfMeasurement::UOM_KM);
$unitOfMeasurement->setDescription('Kilometers');
$locatorRequest->setUnitOfMeasurement($unitOfMeasurement);

try {
    // Get the locations
    $locator = new Ups\Locator($accessKey, $userId, $password);
    $locations = $locator->getLocations($locatorRequest, \Ups\Locator::OPTION_UPS_ACCESS_POINT_LOCATIONS);

	foreach($locations->SearchResults->DropLocation as $location) {
		// Your code here
		var_dump($location);
	}

} catch (Exception $e) {
	var_dump($e);
}
```

<a name="locator-class-parameters"></a>
### Parameters

Locator class parameters are:

 * `locatorRequest` Mandatory. locatorRequest object with request details, see example
 * `requestOption` Optional. Type of locations you are searching for.

<a name="tradeability-class"></a>
## Tradeability Class

The Tradeability class allows you to get data for international shipments:
* Landed Costs (e.g. duties)
* Denied Party Screener
* Import Compliance
* Export License Detection

Note: only the Landed Costs API is currently implemented.

⚠️ Tradeability is only available through a SOAP API. Therefore you are required to have the [SOAP extension](http://php.net/manual/en/book.soap.php) installed on your system.

<a name="tradeability-class-example"></a>
### Example

```php
// Build request
$landedCostRequest = new \Ups\Entity\Tradeability\LandedCostRequest;

// Build shipment
$shipment = new \Ups\Entity\Tradeability\Shipment;
$shipment->setOriginCountryCode('NL');
$shipment->setDestinationCountryCode('US');
$shipment->setDestinationStateProvinceCode('TX');
$shipment->setResultCurrencyCode('EUR');
$shipment->setTariffCodeAlert(1);
$shipment->setTransportationMode(\Ups\Entity\Tradeability\Shipment::TRANSPORT_MODE_AIR);
$shipment->setTransactionReferenceId('1');

// Build product
$product = new \Ups\Entity\Tradeability\Product;
$product->setProductName('Test');
$tariffInfo = new \Ups\Entity\Tradeability\TariffInfo;
$tariffInfo->setTariffCode('5109.90.80.00');
$product->setTariffInfo($tariffInfo);
$product->setProductCountryCodeOfOrigin('BD');
$unitPrice = new \Ups\Entity\Tradeability\UnitPrice;
$unitPrice->setMonetaryValue(250);
$unitPrice->setCurrencyCode('EUR');
$product->setUnitPrice($unitPrice);
$weight = new Ups\Entity\Tradeability\Weight;
$weight->setValue(0.83);
$unitOfMeasurement = new \Ups\Entity\Tradeability\UnitOfMeasurement;
$unitOfMeasurement->setCode('kg');
$weight->setUnitOfMeasurement($unitOfMeasurement);
$product->setWeight($weight);
$quantity = new \Ups\Entity\Tradeability\Quantity;
$quantity->setValue(5);
$unitOfMeasurement = new \Ups\Entity\Tradeability\UnitOfMeasurement;
$unitOfMeasurement->setCode(\Ups\Entity\Tradeability\UnitOfMeasurement::PROD_PIECES);
$quantity->setUnitOfMeasurement($unitOfMeasurement);
$product->setQuantity($quantity);
$product->setTariffCodeAlert(1);

// Add product to shipment
$shipment->addProduct($product);

// Query request
$queryRequest = new \Ups\Entity\Tradeability\QueryRequest;
$queryRequest->setShipment($shipment);
$queryRequest->setSuppressQuestionIndicator(true);

// Build
$landedCostRequest->setQueryRequest($queryRequest);

try {
    // Get the data

CHANGELOG.md
# CHANGELOG

## 0.9.5 (released 05-10-2021)
- fixed LabelSpecification

## 0.9.4 (released 15-07-2020)
- pickup type is optional

## 0.9.3 (released 18-05-2020)
- fixed warning on an empty api response 

## 0.9.2 (released 04-05-2020)
- fixed warning on an empty api response 

## 0.9.1 (released 16-01-2020)

- changed log messages 

## 0.7.12 (released 19-03-2016)

- Extend ShipmentServiceOptions in Shipping API
- Fix for Begin and End date in Tracking API
- Support for HazMat containers in Shipping API 

## 0.7.11 (released 15-09-2016)

- Multiple results in Tracking + add Simple Address Validation class

## 0.7.10 (released 12-07-2016) 

- Solve an issue when mb_detect_encoding couldn't detect it, use a fallback to use utf8_encode. Else response parsing breaks.

## 0.7.9 (released 01-07-2016)

- Add Shipping API Support for AdditionalDocumentIndicator in the InternationalForms node
- Add Shipping API Support for EEIFilingOption in the InternationalForms node
- Add TimeInTransit API Response Service Code Constants for US/EU Shipments to Entity/Service.php
- Add Tracking API Response StatusType Constants

## 0.7.8 (released 23-06-2016)

- Do not create new Guzzle object instance on each Request, but re-use it. 

## 0.7.7 (released 31-05-2016)

- Added AccessPointCOD under ShipmentServiceOptions
- RatedShipment extended with extra fields
- Password field (under Access tag) support entities now

## 0.7.6 (released 04-03-2016)

- Add extra parameters for filtering on Tracking API

## 0.7.5 (released 01-03-2016)

- Improved Address Validation returned object

## 0.7.4 (released 21-02-2016)

- Bugfix: switched node names in XAV

## 0.7.3 (released 17-02-2016)

- Mail Innovation support in Tracking    
- Option to get a result object from the Validation class with several methods to make you process the results easier. Does not introduce backwords incompatibility, as it's an optional feature.

## 0.7.2 (released 09-01-2016)

- Bugfix: Use SoapRequest instead of Request in Tradeability

## 0.7.1 (released 23-11-2015)

- Adds support for a second reference number

## 0.7.0 (released 16-11-2015)

- **[!]** Default ShipFrom on Shipment class not set anymore in constructor (ShipFrom is optional)
- Adds support for cash on delivery for shipments

Items marked with **[!]**  may incur backwards incompatibility.

## 0.6.3 (released 10-11-2015)

- Improvement in parsing XML

## 0.6.2 (released 10-11-2015)

- Add Landed Cost request of Tradeability API (using SOAP). Tradeability consist of 4 endpoints, of which now one is implemented.

## 0.6.1 (released 30-10-2015)

- Add option to use the Tracking API also when supplying a ReferenceNumber

## 0.6.0 (released 25-09-2015)

- Extra check on response in QuantumView, when no response it gave an error
- Added ShipmentRequestLabelSpecification class for easier options setting
- Added ShipmentRequestReceiptSpecification class for easier options setting
- **[!]** Shipment class dropped some public properties in favor of private properties and setter/getter methods.
- **[!]** `confirm` and `accept` methods of Shipping class now receive Shipment, ShipmentRequestLabelSpecification and
ShipmentRequestReceiptSpecification

Items marked with **[!]**  may and will incur backwards incompatibility.

## 0.5.2 (released 16-09-2015)

- TimeInTransit ServiceSummary results should be array of summaries, which was not the case when 1 result

## 0.5.1 (released 08-09-2015)

- Limit alternate delivery address names to 35 characters

## 0.5.0 (released 26-08-2015)

- Added UTF8 compatibility for UPS responses
- Added Guzzle to handle requests
- Changed required PHP version to 5.5
- Removed Autoloader in favor of composer


src/Tracking.php
<?php

namespace Ups;

use DOMDocument;
use Exception;
use Psr\Log\LoggerInterface;
use SimpleXMLElement;
use stdClass;
use DateTime;

/**
 * Tracking API Wrapper.
 */
class Tracking extends Ups
{
    const ENDPOINT = '/Track';

    /**
     * @var RequestInterface
     */
    private $request;

    /**
     *
     * Workaround flag to handle Multiple shipment nodes in tracking response
     * See GitHub Issue #117
     *
     * @todo: fix in next major release
     *
     * @var boolean
     */
    protected $allowMultipleShipments = false;

    /**
     * @todo: make private
     *
     * @var ResponseInterface
     */
    public $response;

    /**
     * @var string
     */
    private $trackingNumber;

    /**
     * @var string
     */
    private $referenceNumber;

    /**
     * @var string
     */
    private $requestOption;

    /**
     * @var string
     */
    private $shipperNumber;

    /**
     * @var \DateTime
     */
    private $beginDate;

    /**
     * @var \DateTime
     */
    private $endDate;

    /**
     * @param string|null $accessKey UPS License Access Key
     * @param string|null $userId UPS User ID
     * @param string|null $password UPS User Password
     * @param bool $useIntegration Determine if we should use production or CIE URLs.
     * @param RequestInterface|null $request
     * @param LoggerInterface|null $logger PSR3 compatible logger (optional)
     */
    public function __construct($accessKey = null, $userId = null, $password = null, $useIntegration = false, RequestInterface $request = null, LoggerInterface $logger = null)
    {
        if (null !== $request) {
            $this->setRequest($request);
        }
        parent::__construct($accessKey, $userId, $password, $useIntegration, $logger);
    }

    /**
     * Get package tracking information.
     *
     * @param string $trackingNumber The package's tracking number.
     * @param string $requestOption Optional processing. For Mail Innovations the only valid options are Last Activity and All activity.
     *
     * @throws Exception
     *
     * @return stdClass
     */
    public function track($trackingNumber, $requestOption = 'activity')
    {
        $this->trackingNumber = $trackingNumber;
        $this->requestOption = $requestOption;

        return $this->getFormattedResponse();
    }

    /**
     * Get package tracking information.
     *
     * @param string $referenceNumber Reference numbers can be a purchase order number, job number, etc. Reference number can be added when creating a shipment.
     * @param string $requestOption
     *
     * @throws Exception
     *
     * @return stdClass
     */
    public function trackByReference($referenceNumber, $requestOption = 'activity')
    {
        $this->referenceNumber = $referenceNumber;
        $this->requestOption = $requestOption;

        return $this->getFormattedResponse();
    }

    /**
     * Set shipper number
     *
     * @param string $shipperNumber
     *
     */
    public function setShipperNumber($shipperNumber)
    {
        $this->shipperNumber = $shipperNumber;
    }

    /**
     * Set begin date
     *
     * @param DateTime $beginDate
     *
     */
    public function setBeginDate(DateTime $beginDate)
    {
        $this->beginDate = $beginDate;
    }

    /**
     * Set end date
     *
     * @param DateTime $endDate
     *
     */
    public function setEndDate(DateTime $endDate)
    {
        $this->endDate = $endDate;
    }

    /**
     * @return stdClass
     * @throws Exception
     */
    private function getFormattedResponse()
    {
        $this->response = $this->getRequest()->request(
            $this->createAccess(),
            $this->createRequest(),
            $this->compileEndpointUrl(self::ENDPOINT)
        );
        $response = $this->response->getResponse();

        if (null === $response) {
            throw new Exception('Failure (0): Unknown error', 0);
        }

        if ($response instanceof SimpleXMLElement && $response->Response->ResponseStatusCode == 0) {
            throw new Exception(
                "Failure ({$response->Response->Error->ErrorSeverity}): {$response->Response->Error->ErrorDescription}",
                (int)$response->Response->Error->ErrorCode
            );
        }

        return $this->formatResponse($response);
    }

    /**
     * Check if tracking number is for mail innovations.
     *
     * @return bool
     */
    private function isMailInnovations()
    {
        $patterns = [

            // UPS Mail Innovations tracking numbers
            '/^MI\d{6}\d{1,22}$/',// MI 000000 00000000+

            // USPS - Certified Mail
            '/^94071\d{17}$/',    // 9407 1000 0000 0000 0000 00
            '/^7\d{19}$/',        // 7000 0000 0000 0000 0000

            // USPS - Collect on Delivery
            '/^93033\d{17}$/',    // 9303 3000 0000 0000 0000 00
            '/^M\d{9}$/',         // M000 0000 00

            // USPS - Global Express Guaranteed
            '/^82\d{10}$/',       // 82 000 000 00

            // USPS - Priority Mail Express International
            '/^EC\d{9}US$/',      // EC 000 000 000 US

            // USPS Innovations Expedited
            '/^927\d{23}$/',      // 9270 8900 8900 8900 8900 8900 00

            // USPS - Priority Mail Express
            '/^927\d{19}$/',      // 9270 1000 0000 0000 0000 00
            '/^EA\d{9}US$/',      // EA 000 000 000 US

            // USPS - Priority Mail International
            '/^CP\d{9}US$/',      // CP 000 000 000 US

            // USPS - Priority Mail
            '/^92055\d{17}$/',    // 9205 5000 0000 0000 0000 00
            '/^14\d{18}$/',       // 1400 0000 0000 0000 0000

            // USPS - Registered Mail
            '/^92088\d{17}$/',    // 9208 8000 0000 0000 0000 00
            '/^RA\d{9}US$/',      // RA 000 000 000 US

            // USPS - Signature Confirmation
            '/^9202\d{16}US$/',   // 9202 1000 0000 0000 0000 00
            '/^23\d{16}US$/',     // 2300 0000 0000 0000 0000

            // USPS - Tracking
            '/^94\d{20}$/',       // 9400 1000 0000 0000 0000 00
            '/^03\d{18}$/'        // 0300 0000 0000 0000 0000
        ];

        foreach ($patterns as $pattern) {
            if (preg_match($pattern, $this->trackingNumber)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Create the Tracking request.
     *
     * @return string
     */
    private function createRequest()
    {
        $xml = new DOMDocument();
        $xml->formatOutput = true;

        $trackRequest = $xml->appendChild($xml->createElement('TrackRequest'));
        $trackRequest->setAttribute('xml:lang', 'en-US');

        $request = $trackRequest->appendChild($xml->createElement('Request'));

        $node = $xml->importNode($this->createTransactionNode(), true);
        $request->appendChild($node);

        $request->appendChild($xml->createElement('RequestAction', 'Track'));

        if (null !== $this->requestOption) {
            $request->appendChild($xml->createElement('RequestOption', $this->requestOption));
        }

        if (null !== $this->trackingNumber) {
            $trackRequest->appendChild($xml->createElement('TrackingNumber', $this->trackingNumber));
        }

        if ($this->isMailInnovations()) {
            $trackRequest->appendChild($xml->createElement('IncludeMailInnovationIndicator'));
        }

        if (null !== $this->referenceNumber) {
            $trackRequest->appendChild($xml->createElement('ReferenceNumber'))->appendChild($xml->createElement('Value', $this->referenceNumber));
        }

        if (null !== $this->shipperNumber) {
            $trackRequest->appendChild($xml->createElement('ShipperNumber', $this->shipperNumber));
        }

        if (null !== $this->beginDate || null !== $this->endDate) {
            $DateRange = $xml->createElement('PickupDateRange');

            if (null !== $this->beginDate) {
                $beginDate = $this->beginDate->format('Ymd');
                $DateRange->appendChild($xml->createElement('BeginDate', $beginDate));
            }

            if (null !== $this->endDate) {
                $endDate = $this->endDate->format('Ymd');
                $DateRange->appendChild($xml->createElement('EndDate', $endDate));
            }

            $trackRequest->appendChild($DateRange);
        }

        return $xml->saveXML();
    }

    /**
     * Format the response.
     *
     * @param SimpleXMLElement $response
     *
     * @return stdClass
     */
    private function formatResponse(SimpleXMLElement $response)
    {
        if ($this->allowMultipleShipments) {
            $response = $this->convertXmlObject($response);
            if (!is_array($response->Shipment)) {
                $response->Shipment = [$response->Shipment];
            }
            return $response;
        }

        return $this->convertXmlObject($response->Shipment);
    }

    /**
     * @return RequestInterface
     */
    public function getRequest()
    {
        if (null === $this->request) {
            $this->request = new Request($this->logger);
        }

        return $this->request;
    }

    /**
     * @param RequestInterface $request
     *
     * @return $this
     */
    public function setRequest(RequestInterface $request)
    {
        $this->request = $request;

        return $this;
    }

    /**
     * @return ResponseInterface
     */
    public function getResponse()
    {
        return $this->response;
    }

    /**
     * @param ResponseInterface $response
     *
     * @return $this
     */
    public function setResponse(ResponseInterface $response)
    {
        $this->response = $response;

        return $this;
    }

    /**
     * @param bool $value
     * @return $this
     */
    public function allowMultipleShipments($value = true)
    {
        $this->allowMultipleShipments = $value ? true : false;

        return $this;
    }
}
